### ISSUE TITLE
These two images are more similar than 0.99?

### ISSUE THREAD

**Issue by andreabisello** [rainmanwy/robotframework-SikuliLibrary#55]
in my test case this comparison return TRUE (Screen Should Contain) with min similarity 0.99
![20-contents-read write share-folder-read write share_content_preview_bk](https://user-images.githubusercontent.com/2149815/33388362-1253ca40-d530-11e7-8920-0acf248ead5d.png)
![20-contents-read write share-folder-read write share_content_preview](https://user-images.githubusercontent.com/2149815/33388363-12708efa-d530-11e7-961d-11101dd6d988.png)

and this are the difference

![image](https://user-images.githubusercontent.com/2149815/33388418-3494e198-d530-11e7-91f3-0f8d8340790c.png)

calculated with Online Image Diff.

so the "red area" (the different) are less than 1% of the image?

thanks.

**Comment by RaiMan** [comment]
this is a problem of the current SikuliX version used here, which is 1.1.1.

Since the beginning of Sikuli, the match score was stripped to 2 decimals.

In your case SikuliX will surely evaluate different scores for both images internally (something like 0.999996 and 0.999912). But both are stripped to 0.99.
Another problem of the current implementation is, that every match score > 0.99 is taken as "exact" match.

To tackle those cases in the moment, you have to use SikuliX features, that return the match scores as is. A function doing this is findAll(). But then you have to interpret the result on your own, since it returns an Iterator<Match>. 

Conclusion: for the time being you have to live with this "grey zone"

**Comment by RaiMan** [comment]
... and this will remain as is with SikuliX 1.x
Only SikuliX 2 next year will be more accurate here (upward compatible with version 1 ;-)

**Comment by abioneperhobby** [comment]
thanks @RaiMan . 

making a screenshot comparison searching for UX regression (in this use case, two buttons disappeared) in entire screen is hard to obtain.

i think i can split screen in a matrix, for example a 3x3 , obtaining 9 images .
 
![image](https://user-images.githubusercontent.com/2149815/33420663-b471a35e-d5af-11e7-894b-3bd6b9983b6d.png)

now, a 

"Screen Should Contain"

should be able to detect differences in top left and top center images ?

**Comment by RaiMan** [comment]
The internally used OpenCV matchTemplate() feature returns a statistical average about the equality of all pixels in the image compared to the image searched in (capture of Screen/Region in our case).
Hence the smaller the amount of pixels, that are different, out of all pixels the smaller the difference in the match scores (as in your example).
So a suitable strategy is to restrict the image to what is expected to be different.

In your case you might first find the whole frame and then check only the upper bar of it.

... but I do not know wether this is possible with this framework.

Of course your suggested approach might also work.

**Comment by abioneperhobby** [comment]
@RaiMan at the beginning, i manually create screenshot of the area i aspect can be different, but this approach needs a heavy manual work on different browser, operative system, graphical card drivers, and so on. so i choose to first create a baseline of screenshot in a release version of the product i know works, and then check agains these auto generated screenshots.

to generate screenshot i use Capture Page Screenshot of SeleniumLibrary and to check against screenshot i use Page Should Contain of Sikuli because the Capture Screen of this library cannot accept a filename and my screenshot filename is related to the test name. Now @rainmanwy patched the Capture Screenshot, it return the screenshot name, so eventually i can rename the file returned to obtain the same result, but this is not the problem.

To avoid manually creation of the screenshot, i create a baseline of screenshot with SeleniumLibrary with Capture Page Screenshot -> Filename.

So, i will try to obtain something to create screenshot of an area of the screen, in order to split screen in different areas and then with sikuli make A-B test.

**Comment by RaiMan** [comment]
@abioneperhobby Agreed: in your situation using this framework your approach might solve your problem. Hope you get it.

Version 2 will have a feature similar to the above mentioned Online Image Diff. Something like that is currently already implemented in SikuliX's observer feature with onChange(), that allows to detect and locate changes on the screen. ... but both do not help in your current situation. all the best.

**Comment by abioneperhobby** [comment]
with this
https://github.com/samdobson/image_slicer
i created a keyword that split the image obtained with Capture Page Screenshot in 9 images,
then my
"User Must To See"
now search for the image matrix

	Screen Should Contain		${WHAT}_01_01.png
	Screen Should Contain		${WHAT}_01_02.png
	Screen Should Contain		${WHAT}_01_03.png
	Screen Should Contain		${WHAT}_02_01.png
	Screen Should Contain		${WHAT}_02_02.png
	Screen Should Contain		${WHAT}_02_03.png
	Screen Should Contain		${WHAT}_03_01.png
	Screen Should Contain		${WHAT}_03_02.png
	Screen Should Contain		${WHAT}_03_03.png

so this image, for example

![20-contents-read write share-folder-read write share_content_editing](https://user-images.githubusercontent.com/2149815/33432373-01a49dd4-d5d8-11e7-96ca-5776913384f9.png)

becomes this matrix of images

![20-contents-read write share-folder-read write share_content_editing_02_01](https://user-images.githubusercontent.com/2149815/33432393-0932d0c0-d5d8-11e7-9f62-f76fbee51514.png)
![20-contents-read write share-folder-read write share_content_editing_02_02](https://user-images.githubusercontent.com/2149815/33432395-0954aae2-d5d8-11e7-8df0-848090b97612.png)
![20-contents-read write share-folder-read write share_content_editing_02_03](https://user-images.githubusercontent.com/2149815/33432396-098526f4-d5d8-11e7-91ab-a8d3002c3824.png)
![20-contents-read write share-folder-read write share_content_editing_03_01](https://user-images.githubusercontent.com/2149815/33432398-09c08ffa-d5d8-11e7-9273-6275d41a36c0.png)
![20-contents-read write share-folder-read write share_content_editing_03_02](https://user-images.githubusercontent.com/2149815/33432399-09e09d4a-d5d8-11e7-9c30-155117fa1d37.png)
![20-contents-read write share-folder-read write share_content_editing_03_03](https://user-images.githubusercontent.com/2149815/33432400-09fc21fa-d5d8-11e7-87b8-75ce6d6819d0.png)
![20-contents-read write share-folder-read write share_content_editing_01_01](https://user-images.githubusercontent.com/2149815/33432401-0a237200-d5d8-11e7-8afa-f2f9a549f0cc.png)
![20-contents-read write share-folder-read write share_content_editing_01_02](https://user-images.githubusercontent.com/2149815/33432402-0a4b06a8-d5d8-11e7-9537-ddef429bf331.png)
![20-contents-read write share-folder-read write share_content_editing_01_03](https://user-images.githubusercontent.com/2149815/33432403-0a69665c-d5d8-11e7-820b-04159c52e55c.png)

so i hope it can works better.

it it works better, i will make a better keyword

**Comment by rainmanwy** [comment]
@abioneperhobby , so you could split the captured image to several images.  As we discussed in issue #54 ,  whether these keywords are needed for this situation?

**Comment by abioneperhobby** [comment]
@rainmanwy i don't know if sikuli can obtain benefit searching an image in a smaller region than the screen region. if yes, setting area can be useful for sikuli. if there are no different, the keyword are not needed anymore (even i solved the problem as a workaround).

**Comment by RaiMan** [comment]
@abioneperhobby 

> i don't know if sikuli can obtain benefit searching an image in a smaller region than the screen region

Benefits are to avoid false positives, if the image is visible more than once, and that the search is faster the smaller the search region.

**Comment by abioneperhobby** [comment]
so i think now we can close this issue and we can improve documentation with these points 

1) match score > 0.99 is taken as exact
2) scores are stripped to two decimal points so 0.99x and 0.99y are considered the same : 0.99

what do you think?

**Comment by RaiMan** [comment]
I can only speak her for SikuliX:

> 1. match score > 0.99 is taken as exact

this is correct

> 2. scores are stripped to two decimal points so 0.99x and 0.99y are considered the same : 0.99

not all over the place and not with all features

example

```
match = find(Pattern(someImage).exact())
match.getScore()

# will print something like
0.999995784
# hence not stripped here
```
